File: SeLeAs/LeetCode/LeetCode 75/deleteNode.py
class Solution:
    def helper(self, root):
        if root.left is None:
            return root.right
        if root.right is None:
            return root.left
        left = root.left
        right = root.right

        # Find the rightmost node in the left subtree
        temp = left
        while temp.right:
            temp = temp.right
        temp.right = right  # Attach right subtree to rightmost node of left
        return left

    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root:
            return None
        if root.val == key:
            return self.helper(root)

        curr = root
        while root:
            if key < root.val:
                if root.left and root.left.val == key:
                    root.left = self.helper(root.left)
                    break
                root = root.left
            else:
                if root.right and root.right.val == key:
                    root.right = self.helper(root.right)
                    break
                root = root.right
        return curr  # Move return outside the loop

File: Self_Learning_Assignments/LeetCode/LeetCode_75/deleteNode.java

public class deleteNode{
class Solution {
    private TreeNode helper(TreeNode root) {
        if (root.left == null) {
            return root.right;
        }
        if (root.right == null) {
            return root.left;
        }

        TreeNode left = root.left;
        TreeNode right = root.right;

        // Find the rightmost node in the left subtree
        TreeNode temp = left;
        while (temp.right != null) {
            temp = temp.right;
        }
        temp.right = right;  // Attach right subtree to the rightmost node of left
        return left;
    }

    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) {
            return null;
        }
        if (root.val == key) {
            return helper(root);
        }

        TreeNode curr = root;
        while (root != null) {
            if (key < root.val) {
                if (root.left != null && root.left.val == key) {
                    root.left = helper(root.left);
                    break;
                }
                root = root.left;
            } else {
                if (root.right != null && root.right.val == key) {
                    root.right = helper(root.right);
                    break;
                }
                root = root.right;
            }
        }
        return curr;
    }
}
}